src/protocol/malicious.rs

```diff
@@ -0,0 +1,306 @@
+            internally_indexed_ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        let (a_left, a_right) = a.as_tuple();
+        let (b_left, b_right) = b.as_tuple();
+        (a_left + a_right) * (b_left + b_right) - a_right * b_right
+    }
+
+    #[allow(dead_code)]
+    fn accumulate_macs(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+        rx: Replicated<F>,
+    ) {
+        // The helpers need to use the same shared randomness to generate the random constant used to validate a given multiplication.
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let prss = ctx.narrow(&"validate".to_string()).prss();
+        let random_constant = prss.generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
+        self.w += Self::compute_dot_product_contribution(random_constant, x);
+    }
+
+    ///
+    /// For each input, `x` of the arithmetic circuit, call this method to obtain
+    /// a secret-sharing of `r * x`.
+    /// This will also affect the information theoretic MAC that is eventually validated
+    /// once the circuit is complete and before outputs are revealed
+    ///
+    #[allow(dead_code)]
+    pub async fn input(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+    ) -> Result<Replicated<F>, BoxError> {
+        let rx = ctx
+            .narrow(&"validate_input".to_string())
+            .multiply(record_id)
+            .await
+            .execute(self.r_share, x)
+            .await?;
+
+        self.accumulate_macs(ctx, record_id, x, rx);
+        Ok(rx)
+    }
+
+    ///
+    /// Each time a multiplication is performed, it must be performed twice, in order to compute both the desired result: `z`,
+    /// as well as the result `r * z`. Once these two values are obtained, they should be passed to this function.
+    #[allow(dead_code)]
+    pub fn validate_multiplication_output(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        z: Replicated<F>,
+        rz: Replicated<F>,
+    ) {
+        self.accumulate_macs(ctx, record_id, z, rz);
+    }
+
+    #[allow(dead_code)]
+    pub async fn validate(&self) -> Result<bool, BoxError> {
```

Comment:
btw, I just noticed that this function returns `bool`, does that mean that protocols will have to validate the value to be `true`? I guess the behavior will be the same (abort the whole process) so I wonder if we can make `validate` implement that (i.e. return `bool` and panic instead)? 

Btw it is important to ensure that we can reliably abort the whole process, by default I believe `panic` just terminates the thread so in some cases we may end up running query even after validator/protocol reported an issue.

We should probably talk to Martin or Andy about this, I believe we can either call `process::abort` (which is not possible to test imo) or put `panic=abort` in `Cargo.toml`. It would be great to understand how we are going to deal with it, regardless of the exact place we decide to put the validation check

Reply:
I think that this shouldn't panic, but should instead return `Result<(), BoxError>` with a validation failure producing the `Err` arm.

Generally speaking, we should ensure that `Result::Err` causes processing to cease.

Reply:
OK, makes sense. I'll update the return type.